src/cpp/session/modules/SessionCodeTools.R

```diff
@@ -2071,3 +2077,148 @@
    descPath <- file.path(packagePath, "DESCRIPTION")
    read.dcf(descPath, all = TRUE)
 })
+
+.rs.addFunction("readSourceDocument", function(id)
+{
+   contents <- .Call("rs_readSourceDocument", as.character(id), PACKAGE = "(embedding)")
+   Encoding(contents) <- "UTF-8"
+   contents
+})
+
+.rs.addFunction("discoverPackageDependencies", function(id, type)
+{
+   # check to see if we have available packages -- if none, bail
+   available <- .rs.availablePackages()
```

Comment:
It seems like the very first call to discoverPackageDependencies will trigger the package list to be cached but this call (and any subsequent ones made prior to the completion of available.packages) will be dropped on the floor. Do you think it'd be worth emitting a client event when we've completed the cache population so that we can finish checking for dependencies? 

I'm imagining e.g. a scenario wherein the IDE boots into a session with R files already open (e.g. as part of a first-use project). Will those get package install prompts?